text/0000-raw-reference-operator.md

```diff
@@ -0,0 +1,182 @@
+In particular, references must be aligned and dereferencable, even when they are
+created and never used.
+
+One consequence of these rules is that it becomes essentially impossible to
+create a raw pointer pointing to an unaligned struct field: `&packed.field as
+*const _` creates an immediate unaligned reference, triggering undefined
+behavior because it is not aligned.  Similarly, `&(*raw).field as *const _` is
+not just computing an offset of the raw pointer `raw`, it also asserts that the
+intermediate shared reference is aligned and dereferencable.  In both cases,
+that is likely not what the author of the code intended.
+
+To fix this, we propose to introduce a new primitive operation on the MIR level
+that, in a single MIR statement, creates a raw pointer to a given place.  No
+intermediate reference exists, so no invariants have to be adhered to.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When working with unaligned or potentially dangling pointers, it is crucial that
+you always use raw pointers and not references: References come with guarantees
+that the compiler assumes are always upheld, and these guarantees include proper
+alignment and not being dangling.  Importantly, these guarantees must be
+maintained even when the reference is created and never used!  The following is
+UB (assuming `packed` is a variable of packed type):
+
+```rust
+let x = unsafe { &packed.field }; // `x` is not aligned -> undefined behavior
+```
+
+There is no situation in which the above code is correct, and hence it is a hard
+error to write this.  This applies to most ways of creating a reference, i.e.,
+all of the following are UB if `X` is not properly aligned and dereferencable:
+
+```rust
+fn foo() -> &T {
+  &X
+}
+
+fn bar(x: &T) {}
+bar(&X); // this is UB at the call site, not in `bar`
+
+let &x = &X; // this is actually dereferencing the pointer, certainly UB
+let _ = &X; // throwing away the value immediately changes nothing
+&X; // different syntax for the same thing
+
+let x = &X as &T as *const T; // this is casting to raw "too late"
+```
+
+The only way to create a pointer to an unaligned or dangling location without
+triggering undefined behavior is to *immediately* turn it into a raw pointer.
+All of the following are valid:
+
+```rust
+let x = unsafe { &packed.field as *const _ };
+let y: *const _ = unsafe { &packed.field };
+```
+
+The intention is to cover all cases where a reference, just created, is
+immediately explicitly used as a value of raw pointer type.
+
+These two operations (taking a reference, casting/coercing to a raw pointer) are
+actually considered a single operation happening in one step, and hence the
+invariants incurred by references do not come into play.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When translating HIR to MIR, we recognize `&[mut] <place> as *[mut|const] _` as
```

Comment:
Can you clarify whether `*[mut|const] _` must literally have `_` or if any `T` is OK?
E.g. is it OK to do `&place as *const u8`?

Reply:
I assume `_` is a meta-inference-variable and the pointee type doesn't matter.

Reply:
@eddyb That's my inference as well; but it should be more clearly specified. :)

Reply:
I updated the text.